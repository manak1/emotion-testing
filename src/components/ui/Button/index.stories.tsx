import { ComponentStory, ComponentMeta } from "@storybook/react"

import Component from "./index"

// More on default export: https://storybook.js.org/docs/react/writing-stories/introduction#default-export
export default {
  title: "ui/Button",
  component: Component,
  argTypes: {
    label: { control: "text", defaultValue: "ボタンだよ" },
    isFullWidth: { control: "boolean", defaultValue: false },
  },
} as ComponentMeta<typeof Component>

// More on component templates: https://storybook.js.org/docs/react/writing-stories/introduction#using-args
const Template: ComponentStory<typeof Component> = (args) => {
  return <Component {...args} />
}

export const Primary = Template.bind({})
Primary.args = {
  label: "ボタン",
}
export const Danger = Template.bind({})
Danger.args = {
  label: "ボタン",
  variant: "danger",
}
export const Cancel = Template.bind({})
Cancel.args = {
  label: "ボタン",
  variant: "cancel",
}
