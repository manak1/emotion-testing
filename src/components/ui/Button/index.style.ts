import { Theme } from "@emotion/react"
import styled from "@emotion/styled"

import { ButtonProps, ButtonVariant } from "./index"

export const Button = styled.button<Omit<ButtonProps, "label">>(
  {
    minWidth: "36px",
    outline: "none",
    padding: "8px 16px",
    border: "none",
    borderRadius: "9999px",
    cursor: "pointer",
  },
  ({ variant, isFullWidth, theme, disabled }) => ({
    fontSize: theme.fontSizes.default,
    width: isFullWidth ? "100%" : "auto",
    fontWeight: "bold",
    "&:hover": {
      filter: "brightness(95%)",
      transition: "filter 0.2s ease-out",
    },
    ...setVariant(variant ?? "primary", theme, disabled),
    ...theme.focus(),
    transition: "box-shadow, filter, 0.2s ease-in",
  })
)

function setVariant(variant: ButtonVariant, theme: Theme, disabled?: boolean) {
  if (disabled) {
    return {
      opacity: 0.7,
      color: "rgba(0, 0, 0, 0.3)",
      backgroundColor: theme.colors.gray,
      pointerEvent: "none",
    }
  }

  switch (variant) {
    case "primary":
      return {
        backgroundColor: theme.colors.primary,
        color: "#fff",
      }
    case "danger":
      return {
        backgroundColor: theme.colors.error,
        color: "#fff",
      }
    case "cancel":
      return {
        backgroundColor: "#fff",
        color: theme.colors.black,
        border: "1px solid #d3d3d3",
      }
  }
}
