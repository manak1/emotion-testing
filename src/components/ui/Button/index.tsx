import { Twitter, HandThumbsUpFill, Link } from "@emotion-icons/bootstrap"
import { css } from "@emotion/react"
import React, { useCallback } from "react"

import * as Styled from "./index.style"

const buttonIcons = {
  twitter: Twitter,
  thumbsUp: HandThumbsUpFill,
  link: Link,
}

export type ButtonVariant =
  | "primary"
  | "danger"
  | "cancel"
  | "outline"
  | "twitter"
export type ButtonSize = "default" | "small"
export type ButtonIcon = keyof typeof buttonIcons

export type ButtonProps = React.ComponentProps<"button"> & {
  variant?: ButtonVariant
  size?: ButtonSize
  icon?: ButtonIcon
  isFullWidth?: boolean
}

const Button: React.FC<ButtonProps> = (props) => {
  const { children, variant = "primary", icon, ...rest } = props

  const renderIcon = useCallback((icon: ButtonIcon) => {
    const IconComponent = buttonIcons[icon]
    return (
      <IconComponent
        size={24}
        css={css`
          margin-right: 8px;
        `}
      />
    )
  }, [])

  return (
    <Styled.Button {...rest} variant={variant} icon={icon}>
      {icon && renderIcon(icon)}
      <Styled.Text>{children}</Styled.Text>
    </Styled.Button>
  )
}

export default Button
