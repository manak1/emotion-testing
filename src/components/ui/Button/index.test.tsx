import { userEvent } from "@storybook/testing-library"
import { render, screen } from "~/test/test-utils"
import Button from "./index"

describe("Buttonのテスト", () => {
  const onClick = jest.fn()

  it("snapshot", () => {
    const { asFragment } = render(<Button label="ボタン" />)
    expect(asFragment()).toMatchSnapshot()
  })

  it("disabled時にonChangeが呼ばれない", () => {
    render(<Button label="ボタン" onClick={onClick} disabled />)
    userEvent.click(screen.getByRole("button"))
    expect(onClick).not.toBeCalled()
  })

  it("clickするとonChangeが呼ばれる", () => {
    render(<Button label="ボタン" onClick={onClick} />)
    userEvent.click(screen.getByRole("button"))
    expect(onClick).toBeCalledTimes(1)
  })
})
