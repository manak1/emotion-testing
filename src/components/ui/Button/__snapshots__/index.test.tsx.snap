// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Buttonのテスト snapshot 1`] = `
<DocumentFragment>
  .emotion-0 {
  min-width: 36px;
  outline: none;
  padding: 10px 16px;
  border: none;
  border-radius: 8px;
  cursor: pointer;
  font-size: 16px;
  width: auto;
  display: initial;
  -webkit-align-items: normal;
  -webkit-box-align: normal;
  -ms-flex-align: normal;
  align-items: normal;
  font-weight: bold;
  background-color: #7856FF;
  color: #fff;
  -webkit-transition: box-shadow,filter,0.2s ease-in;
  transition: box-shadow,filter,0.2s ease-in;
  padding: 10px 16px;
}

.emotion-0:hover {
  -webkit-filter: brightness(95%);
  filter: brightness(95%);
  -webkit-transition: filter 0.2s ease-out;
  transition: filter 0.2s ease-out;
}

.emotion-0:focus {
  box-shadow: 0 0 0 2px rgba(120, 86, 255, 0.3);
  -webkit-transition: box-shadow 0.2s ease-out;
  transition: box-shadow 0.2s ease-out;
}

.emotion-1 {
  margin: 0;
  padding: 0;
}

<button
    class="emotion-0"
  >
    <p
      class="emotion-1"
    >
      ボタン
    </p>
  </button>
</DocumentFragment>
`;
